### DISCUSSION TITLE
Recycle Bin for Packages

### DISCUSSION THREAD

**Discussion by jqnatividad** [ckan/ckan#7363] [Ideas]
Currently, there is a system-wide Trash where the sysadmin can purge "soft-deleted" packages, groups and organizations.

In a current implementation, there is interest in enabling the following functionality:
* being able to "recover"/"[undelete](https://github.com/ckan/ckan/issues/7212)" a "soft-deleted" package, even if there are constraints on what can be recovered (e.g. only the latest revision can be recovered)
* having the ability to have organization-level trash, where org-admins can recover/purge organization packages
* having the ability to configure an aging policy on the trash cans (i.e. automatically purge packages after N days), and see the age of items in the trash can
* having the ability to Search the trash can or at least have a better interface with pagination/sorting.

* as a "stretch" goal - there is also interest in implementing a "personal" trash can, where a user can recover/purge his/her own datasets

Considerations/Questions:
* How will these actions be captured in the Activity Stream so it behaves more like an audit trail? Especially since most of the packages are private (which is not captured on the existing Activity Stream by default).
* Indexing deleted packages for a better Trash UI?
* Being able to see the status/detailed datastore info of related resources? ([datastore_info](https://github.com/ckan/ckan/pull/5846), https://docs.ckan.org/en/2.9/maintaining/datastore.html#ckanext.datastore.logic.action.datastore_run_trigger?)

**Comment by JatinChimote** [comment]
Another may be not directly related request is if we can have a 'after_*_purge' function which tell about us what got purged. So probably it can be used to tell some other dependent services and system about the purge.

**Comment by wardi** [comment]
Definitely expanding the scope of the original topic, but it's time we talked about

## Soft deletion in CKAN

The biggest issue (in terms of space) with soft-deleted datasets is that it's not clear when to delete external data like uploaded files, datastore tables, views, etc. If we delete them on soft-delete then undeleting will create a dataset with broken links and views. @JatinChimote points out that we don't even have hooks at purge (actual delete) time, so we can't delete then either.

So I guess we just never remove files that have been uploaded or datastore data https://github.com/ckan/ckan/issues/5446 . Never deleting is a safe choice, but our users paying for every MB of server storage can't be happy about it. At a minimum we should have a way to scan for and remove uploaded files and datastore tables and views that are no longer linked (plus one for each cloud/blob storage extension too?).

Another problem with soft deletion is that table space isn't free, and can even create significant performance problems for users doing seemingly normal things: https://github.com/ckan/ckan/issues/7119

We also have the issue of soft deleted records preventing new ones from being created: https://github.com/ckan/ckan/issues/4267 and ugly hacks to work around the issue in some situations: https://github.com/ckan/ckan/issues/2438

These are just a handful of the reported issues, and I'm sure many times the number of users reporting issues instead are quietly frustrated.

### Entities supporting soft deletion

Where have we implemented soft deletion? Well, these are our database tables with a state flag:

- group (table used for organizations as well)
- group_extra (used for individual string:string values in groups/orgs)
- member (used for all sorts of many to many relationships)
- package (all dataset types)
- package_extra (used for individual key:value strings in datasets)
- package_relationship (monstly unmaintained)
- package_tag (many to many relationship has a state but *tag* doesn't)
- resource (separate state from parent package)
- system_info (configuration key:value)
- task_status (background job progress)
- user

And some numbers from one of our sites with row and soft-deleted row counts:

| table | total rows | state=deleted |
| --- | --- | --- |
| group | 349 | 7 |
| group_extra | 3120 | 358 |
| member | 29736 | 12209 |
| package | 23987 | 11347 |
| package_extra | 485726 | 894 |
| package_relationship | 0 | 0 |
| package_tag * | 0 | 0 |
| resource | 90520 | 4355 ( + 22879 **) |
| system_info | 0 | 0 |
| task_status | 10403 | 0 |
| user | 5130 | 528 |

\* tags not used on our site due to performance and multilingual limitations

\** resource state=active but parent package is state=deleted

### Where does soft deletion make sense?

We have occasionally restored an accidentally deleted user or package by setting `state=active`. However I can't imagine ever wanting to restore a package_extra or a membership record (as opposed to simply recreating one). Restoring accidentally deleted resources could be nice, but as a user unless I had the foresight to record the resource id there's no way to do that.

The majority of these soft-deleted records are slowing useful database queries with no benefit at all.

I suggest we remove soft deletion and the state column from all these tables except for group, package and user. For group, package and user let's add a CLI command to purge deleted records (with an optional older-than parameter).

If resources no longer have soft delete then we can safely remove uploads, datastore tables etc. when the resource is deleted *or* when their package is purged. Let's cover these cases with tests because leaking storage makes users unhappy.

Let's formalize and document the expected soft deletion behavior: which entities it applies to and how to use it to restore accidentally deleted items. New `*_undelete` API actions would be nice.

And finally, I really like @jqnatividad 's suggestion. :wastebasket: Trash bins :wastebasket: for everyone, not just sysadmins! (limited to just those three entities, user-friendly and with restore covered by tests)

**Comment by wardi** [comment]
copied my comment to a new issue #7368 for visibility
